zerver/lib/topic.py

```diff
@@ -153,55 +155,68 @@ def update_messages_for_topic_edit(
     edit_history_event: EditHistoryEvent,
     last_edit_time: datetime,
 ) -> List[Message]:
-    propagate_query = Q(
-        recipient_id=old_stream.recipient_id,
+    # Uses index: zerver_message_realm_recipient_upper_subject
+    messages = Message.objects.filter(
+        realm_id=old_stream.realm_id,
+        recipient_id=assert_is_not_none(old_stream.recipient_id),
         subject__iexact=orig_topic_name,
     )
     if propagate_mode == "change_all":
-        propagate_query = propagate_query & ~Q(id=edited_message.id)
+        messages = messages.exclude(id=edited_message.id)
     if propagate_mode == "change_later":
-        propagate_query = propagate_query & Q(id__gt=edited_message.id)
-
-    # Uses index: zerver_message_realm_recipient_upper_subject
-    messages = Message.objects.filter(propagate_query, realm_id=old_stream.realm_id).select_related(
-        *Message.DEFAULT_SELECT_RELATED
-    )
-
-    update_fields = ["edit_history", "last_edit_time"]
+        messages = messages.filter(id__gt=edited_message.id)
 
     if new_stream is not None:
         # If we're moving the messages between streams, only move
         # messages that the acting user can access, so that one cannot
         # gain access to messages through moving them.
         from zerver.lib.message import bulk_access_stream_messages_query
 
-        messages_list = list(bulk_access_stream_messages_query(acting_user, messages, old_stream))
+        messages = bulk_access_stream_messages_query(acting_user, messages, old_stream)
     else:
         # For single-message edits or topic moves within a stream, we
         # allow moving history the user may not have access in order
         # to keep topics together.
-        messages_list = list(messages)
-
-    # The cached ORM objects are not changed by the upcoming
-    # messages.update(), and the remote cache update (done by the
-    # caller) requires the new value, so we manually update the
-    # objects in addition to sending a bulk query to the database.
+        pass
+
+    update_fields: Dict[str, object] = {
+        "last_edit_time": last_edit_time,
+        "edit_history": Cast(
+            Func(
+                Cast(
+                    Func(
+                        F("edit_history"),
+                        Value("[]"),
+                        function="COALESCE",
+                    ),
+                    JSONField(),
+                ),
+                Cast(
+                    Value(orjson.dumps([edit_history_event]).decode()),
+                    JSONField(),
+                ),
+                function="",
+                arg_joiner=" || ",
+            ),
+            TextField(),
+        ),
+    }
```

Comment:
I can't say I know how to read this, but I assume you've tested it carefully.

Reply:
I'll add a comment with its translation.